components/butterfly/protocols/newscast.proto

```diff
@@ -12,6 +12,7 @@ message Election {
   optional uint64 suitability = 4;
   optional Status status = 5;
   repeated string votes = 6;
+  optional string expiration = 7;
```

Comment:
That's wild.  I'm not sure how I missed that there was a native `Timestamp` type in protobufs.  I'll investigate this in more depth.

Reply:
A first stab at implementing this leads to many compilation errors around prost's `Message` trait (which is apparently not implemented for `Timestamp`?) and de/serialization.  I've spent a bit of time trying to resolve these without much luck, so for now, I'm going to shelve this effort since what I have works.

Reply:
So after a bit more reading on this, it looks like in order to make this change, we'd need to manually implement the `Serialize` and `Deserialize` traits from `serde`  for `google.protobuf.Timestamp`.

Implementing `Serialize` isn't bad, but `Deserialize` is definitely a non-trivial affair, and to be honest, I don't see the value in going through all that effort, given that the protobufs themselves are the only place where these types exist as a string.  All the Rust code deals with proper `DateTime` types, and the effort required to convert those types to and from strings is trivial.

Reply:
Thinking a bit more on this, is it actually necessary to have expiration in the protobuf type itself? Does one supervisor relay expirations to another, or does everyone maintain their own expirations? My impression was the latter. If that’s the case, we could keep the same protobufs as before, and add the expiration info only to our local internal types.

If I’m wrong though, kindly ignore me :sweat_smile:

Reply:
Also, as an aside, I think you just need to add `prost-types` to `Cargo.toml` to get `google.protobuf.Timestamp` to work. That's about all I did to get this [working for the event subsystem](https://github.com/habitat-sh/habitat/commit/bc32e2c2048239bad833b80ee5eaac5085489f0e)

Reply:
@christophermaier Right now supervisors relay expirations to each other, so having it in the protobuf is necessary.  Regarding your second suggestion, I did that, but that didn't include `Serialize` and `Deserialize` implementations.

Reply:
@raskchanky Is there a way that expirations could be tracked locally? I thought we'd collectively discussed that at some point (though I may be misremembering).

As for the `Serialize` and `Deserialize` bits... there are a few things to dig into here. I don't think we actually need `Deserialize` do we? I haven't seen where that's used. Also, it turns out that we need `Serialize` because `RumorStore` itself implements `Serialize` because we have a `ServerProxy` for Butterfly to generate JSON for the Supervisor's HTTP gateway (whew!). That seemed odd to me (why are we serializing protobuf messages?).

We _do_ have some internal Rust types that we use elsewhere, and in the Supervisor, but we appear to be using the plain protobuf-derived types in the `RumorStore`s... for size, I guess? :man_shrugging: It's a little difficult to tell, because the code and types here are all rather intertwined in non-obvious ways.

If we really don't need `Deserialize`, and implementing `Serialize` is easy, then it might be a win to use the Google type in the protobufs. Otherwise, we can go with what you've already got here, pending a bigger cleanup.

Reply:
I might be thinking about this wrong, but I don't see how expirations can be tracked locally.  If you have a 3 node supervisor network and you e.g. `hab svc unload core/redis` on supervisor A, and the corresponding service rumors get an expiration date, how do supervisors B and C know about the expiration and forthcoming removal of those service rumors if you're not sending the expiration date in the rumor protobufs?

I'll double check on `Serialize` and `Deserialize` but mostly I'm going off of the compiler errors that I got when I tried to change the type to a `Timestamp`.

Ultimately, while this change would be nice, I don't see that it really nets us anything substantial other than feeling good about the types in our protobufs.  Given that all the Rust code is still type safe, the odds of any malformed values making their way into the protobufs seems unlikely.  It's possible I'm missing something, but I feel like the net gain from this change is pretty small.

Reply:
Using standard types when possible is generally good, and it'd mean that you could dispense with the string parsing. It'd be more efficient across the network, too, though that's probably a tiny benefit.

Thinking a bit more... yeah, I guess we do need to gossip the expiration stamps. However, it just occurred to me that looks like we're basically using them as flags, rather than actual timestamps, right? That is, everything expires at "never", until we expire it, then we tell everyone that the rumor expires "soon". (Am I understanding that correctly?)

If that's the case, we could basically just have a boolean flag, right? If you get a rumor with the flag set to `true`, drop it in an hour.

Is that all we'd need to do to gossip that services have shut down? :thinking:

Reply:
Separate but related thought: how do things behave in the following scenario:

1) Supervisor A stops Service X
2) Supervisor A gossips that Service X is expiring
3) Supervisor B acknowledges that Service X is expiring (in an hour, say)
4) Before the hour is up, Supervisor A starts Service X again.

Does Supervisor B recognize that Service X is no longer expiring?

Reply:
@christophermaier To your first point, yes, that's how it works, and thinking about it now, I guess you could replace the date with a boolean flag.  When this work first started, I had the idea that the individual rumors were going to need to expire at different times, but I think that's proved to not be the case.  Maybe it's easier to just set a boolean flag.  @baumanj - what are your thoughts on going this route?

To your second point, yes, Supervisor B should recognize that Service X is no longer expiring, because Supervisor A will generate a new `Service` rumor, which should replace the existing one via the `merge` function, and the new one will not have an expiration date on it, which will then get gossiped to Supervisor B and replace its existing rumor.  Having a boolean flag feels like it would simplify this scenario as well.

Reply:
I feel that there's benefit in the explicit scoping of expiration versus a boolean. I think there is a case for some classes of rumors to never expire (such as `Departure`), but I'm less sold on whether `Transient` rumors (such as `Service`) should ever have an expiration of "never". I thought part of the original idea here was to move towards a smaller working set of rumors. Unfortunately, we can't rely on the remote nodes to always reliably tell us when they're unavailable, so it seems simpler to assume a rumor is no longer valid if not refreshed after a reasonable period.

Reply:
The point of view that I came around to (which is represented by the PR in its current state) is that the collection of rumor stores and their values make up butterfly's state.  As such, it doesn't make a lot of sense to expire parts of that state or to have to refresh it frequently.  It exists as long as it is required for the job at hand and when it's no longer required (e.g. a `Member` departs, a `Service` is unloaded, etc) then we purge it.  In that sense, I think a boolean flag is a better fit than an expiration date.

In the first iteration of this work (which was on a separate branch), I tried to implement the idea of all rumors having an expiration date and having them expire on their own if they weren't periodically refreshed.  That sounded great in theory, but was fraught with peril in practice, and thinking on it further brought me to the current approach.

In the current state of things, rumors serve a dual purpose of 1) a communication medium and also 2) storing butterfly's state.  In my opinion, this is sub-optimal and could be streamlined further in the future by having a more straightforward way to store butterfly's state and probably a different way to synchronize it between nodes as well.  That's likely out of scope for this PR though.

Reply:
Can you talk more about the the first iteration and why it was fraught with peril?

In its current iteration, I wouldn't really classify this as expiration (as indicated by our consideration of a boolean) so much as adding the potential for negative rumors.

The reason I was originally inclined for us to go with the concept of expiring rumors is that it decouples the operation of the different nodes. What I'm seeing now requires predictable coordination between remote entities. Services have to be unloaded or members need to depart and that has to be gracefully communicated across the network. However, in practice, a node may just stop hearing from another one. If the partition is complete, that's easy enough to deal with, but what if the other node appears to be alive, but it's no longer running the service we believe it was, though we haven't received any positive communication to that effect. Do we operate on old information and assume it's still valid? For how long?

I think one of the really hard parts about distributed systems is that *everything* is so much more error prone, so the kinds of assumptions that one can make within local processes (or well behaved, small networks) tend to fall apart in practice with large, messy networks. Chief among these is the assumption that we'll hear something (or that we'll definitely hear *nothing*, as opposed to echoes of old rumors or something else strange) when something goes wrong. I do imagine there's some additional complexity to moving toward a world where we only believe the state of a remote node is a certain way when we've had a timely affirmation of it, but I think in terms of reasoning about a distributed system, it's ultimately the more robust path.

Reply:
It's been awhile since I looked at that older branch, but my memory of where things started to fall apart was around refreshing rumors - where/when should rumors get refreshed?  How does the refresh mechanism interact with the `merge` mechanism?

I'm not opposed to re-working this a bit if we want to go back to that kind of scheme.  Maybe there was something I missed the first time around.

Reply:
Just to follow up on this - I'm going to take another stab at the refresh mechanism that I couldn't get working a few months ago.